fix(txt-registry): skip creation of already-existing TXT records (#4914)

u-kai opened the pull request:
## What does it do ?

Adds a cache-based filter to **TXTRegistry** so that `External-DNS` no longer
tries to recreate `TXT` records that already exist in the provider.  
**Fixes #4914**

## Motivation

When only the data record (`A/CNAME`) is removed, `External-DNS` next sync loop
recreates that record and also tries to create the unchanged `TXT` record,
which providers like `Route53` reject with “record already exists”.  
Caching existing `TXT` entries and skipping them in `ApplyChanges()` prevents
this failure.

## More

- [x] Yes, this PR title follows [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Yes, I added unit tests
- [ ] Yes, I updated end user documentation accordingly

mloiseleur commented:
Thanks. 
I'm not sure of this approach :thinking: It may increase memory usage a lot on big users. 
/ok-to-test
cc @szuecs

mloiseleur commented:
What happens the txt entry has also been deleted ? 
It will still be in this cache, so not re-created ?

u-kai commented:
@mloiseleur 
 Hi, I noticed that the test failure in this PR might be due to a flaky test.

It seems that the failure is caused by the use of a `reg.MustRegister` with `randInt(100)` in [main_test.go](https://github.com/kubernetes-sigs/external-dns/blob/b0cc39a2aee9d875da49c8f9ea0f7d8a94ce1bf5/internal/gen/docs/metrics/main_test.go#L60-L98), and it looks like we’ve just hit the 1% chance of getting the same random value twice.

To improve this, I’m thinking of replacing `randInt(100)` with something like `randInt(10000)` to reduce the chance of collision.

Would you prefer this fix to be included in this PR, or should I open a separate one for it?

u-kai commented:
@mloiseleur 
Thank you for the suggestion!
I submitted a separate [PR](https://github.com/kubernetes-sigs/external-dns/pull/5462) to address the test flakiness using a different approach.

I'd really appreciate it if you could continue reviewing this PR(fix txt-registry) as well.
Thanks again!

ivankatliarchuk commented:
There are concerns Regarding TXT Caching as a Solution

I'm unclear how caching will resolve the underlying issue. From my perspective, this looks like a bug in either the CRD source, the TXT registry, or how records and plans are being handled. Instead of identifying and fixing that root cause, the current proposal is to introduce a new feature – TXT caching. The issue missing how-to reproduce, what the arguments are and version for reported bug is ~0.13, when latest version is 0.17, which means, the issue may no longer even exist.

The approach could actually worsen the situation for TXT records. Implementing a cache on a stateless service means the problem is likely to reappear during restarts or new service startups.

This proposed caching mechanism isn't a fix. We need to find the root cause of the problem. The expectation should be to reproduce the issue in a controlled environment and then try to pinpoint where the bug lies.

Relevant issue https://github.com/kubernetes-sigs/external-dns/issues/4998

Relevant solution https://github.com/kubernetes-sigs/external-dns/pull/5372

It appears there's a bug within either the CRD source, the TXT registry, or the mechanism for pulling records for a specific provider and handling the plan. Instead of addressing this underlying bug, the decision has been made to introduce a new feature: TXT caching.

ivankatliarchuk commented:
TXT registry already have a cache https://github.com/kubernetes-sigs/external-dns/blob/36bc7d6bc47d4d8862aa02f88ac57d0608919a2a/registry/txt.go#L46 and local endpoints and TXT records cache https://github.com/kubernetes-sigs/external-dns/blob/master/registry/txt.go#L138-L141

This is a high-impact issue, and without being able to reproduce it, it's tough to pinpoint the exact cause. It's even possible that `upsert-only` failed, but everything works fine when the `sync` flag is set.

I propose to start with:

- Adding a unit test that specifically shows the issue where the record is deleted but the registry record persists.
- Reproducing the actual bug on a local cluster using the Route53 provider and the latest external-dns.
- Implementing the fix once the root cause is identified.
- 
First two could be done in any order, but with focus on reproducing the problem and providing a unit test, that reveals the bug.

This is currently just an assumption, and without reproducing it, it's hard to be certain. We need to confirm that if a TXT registry record exists without a corresponding managed record (like A, AAAA, or CNAME), we either:

- Mark the TXT registry record for deletion (though this won't work with upsert-only) and then apply the new changes, OR
- Leave the TXT registry record untouched and only update the actual record.

ivankatliarchuk commented:
Fixes https://github.com/kubernetes-sigs/external-dns/issues/5340

u-kai commented:
Hi @mloiseleur,  I mentioned earlier that I would wait for your confirmation before pushing the changes, but I received a detailed comment from @ivankatliarchuk and pushed the revised implementation to help clarify the current behavior.

The latest code uses a temporary structure between `Records` and `ApplyChanges` instead of caching across reconciliation loops, as previously discussed.

Let me know if this direction still looks good to you — I’m happy to adjust if needed!

ivankatliarchuk commented:
Have you managed to validate that issue is still present or it's related to specific flags
- upsert-only
- sync
?

ivankatliarchuk commented:
relates: https://github.com/kubernetes-sigs/external-dns/issues/3977

u-kai commented:
@ivankatliarchuk 

Yes, I was able to reproduce the issue with both `--policy=sync` and `--policy=upsert-only`.
For clarity, I'm using `hoge.com` here as a placeholder — the actual domain has been replaced.

### Environment
- branch: master
- command: 
   ```
    external-dns \
  --provider aws \
  --registry txt \
  --txt-owner-id demo \
  --policy sync \ # and upsert-only
  --log-level debug \
  --source ingress
   ```
- ingress resource
  ```yaml
  apiVersion: networking.k8s.io/v1
   kind: Ingress
  metadata:
    name: demo-wild
    annotations:
      external-dns.alpha.kubernetes.io/aws-weight: "100"
      external-dns.alpha.kubernetes.io/hostname: api.hoge.com
      external-dns.alpha.kubernetes.io/set-identifier: demo
  spec:
    rules:
      - host: "api.hoge.com"
        http:
          paths:
            - path: /
              pathType: Prefix
              backend:
                service:
                  name: demo-service
                  port:
                    number: 80
  ```

### Steps to Reproduce
1. Deploy ExternalDNS.
1. Deploy the above Ingress and let ExternalDNS create the records.
1. Manually delete the A record for `api.hoge.com` from Route53.

Wait for the next reconciliation loop.

### Result
On the next loop, ExternalDNS tries to recreate the A record and the associated TXT record. 
However, since the TXT record still exists, Route 53 throws an error due to the attempted re-creation of an already-existing record.

```
DEBU[0001] Adding api.hoge.com. to zone hoge.com. [Id: /hostedzone/ZONEIDXXXXXXXXX]
DEBU[0001] Adding api.hoge.com. to zone hoge.com. [Id: /hostedzone/ZONEIDXXXXXXXXX]
DEBU[0001] Adding a-api.hoge.com. to zone hoge.com. [Id: /hostedzone/ZONEIDXXXXXXXXX]
INFO[0001] Desired change: CREATE a-api.hoge.com TXT  profile=default zoneID=/hostedzone/ZONEIDXXXXXXXXX zoneName=hoge.com.
INFO[0001] Desired change: CREATE api.hoge.com A   profile=default zoneID=/hostedzone/ZONEIDXXXXXXXXX zoneName=hoge.com.
INFO[0001] Desired change: CREATE api.hoge.com TXT  profile=default zoneID=/hostedzone/ZONEIDXXXXXXXXX zoneName=hoge.com.
...
DEBU[0061] Adding api.hoge.com. to zone hoge.com. [Id: /hostedzone/ZONEIDXXXXXXXXX]
DEBU[0061] Adding api.hoge.com. to zone hoge.com. [Id: /hostedzone/ZONEIDXXXXXXXXX]
DEBU[0061] Adding a-api.hoge.com. to zone hoge.com. [Id: /hostedzone/ZONEIDXXXXXXXXX]
INFO[0061] Desired change: CREATE a-api.hoge.com TXT  profile=default zoneID=/hostedzone/ZONEIDXXXXXXXXX zoneName=hoge.com.
INFO[0061] Desired change: CREATE api.hoge.com A   profile=default zoneID=/hostedzone/ZONEIDXXXXXXXXX zoneName=hoge.com.
INFO[0061] Desired change: CREATE api.hoge.com TXT  profile=default zoneID=/hostedzone/ZONEIDXXXXXXXXX zoneName=hoge.com.
ERRO[0061] Failure in zone hoge.com. when submitting change batch: operation error Route 53: ChangeResourceRecordSets, https response error StatusCode: 400, RequestID: 08d2eafe-c886-46d3-bd49-c5f905a7d3ee, InvalidChangeBatch: [Tried to create resource record set [name='a-api.hoge.com.', type='TXT', set-identifier='demo'] but it already exists, Tried to create resource record set [name='api.hoge.com.', type='TXT', set-identifier='demo'] but it already exists]  profile=default zoneID=/hostedzone/ZONEIDXXXXXXXXX zoneName=hoge.com.
ERRO[0062] Failed to do run once: soft error
failed to submit all changes for the following zones: [/hostedzone/ZONEIDXXXXXXXXX]
```

ivankatliarchuk commented on registry/txt.go:
```diff
@@ -285,6 +335,8 @@ func (im *TXTRegistry) ApplyChanges(ctx context.Context, changes *plan.Changes)
 			im.addToCache(r)
 		}
 	}
+	filteredChanges.Create = im.existingTXTs.filterOutExistingTXTRecords(filteredChanges.Create)
```
we not necessary need that logic, as this is another loop.

ivankatliarchuk replied:
instead we could have something like

```go
for _, r := range filteredChanges.Create {
		if im.existingTXTs.isManaged(r) {
			continue
		}
...
}
```

ivankatliarchuk commented on registry/txt.go:
```diff
@@ -285,6 +335,8 @@ func (im *TXTRegistry) ApplyChanges(ctx context.Context, changes *plan.Changes)
 		}
 	}
+	filteredChanges.Create = im.existingTXTs.filterOutExistingTXTRecords(filteredChanges.Create)
+	im.existingTXTs = newExistingTXTs() // reset existing TXTs for the next reconciliation cycle
```
```go
im.existingTXTs.clean() or empty()
```

u-kai commented:
@ivankatliarchuk 
Thanks a lot for the review and helpful suggestion!
I've reflected the changes accordingly.

u-kai commented on registry/txt.go:
```diff
@@ -279,12 +318,20 @@ func (im *TXTRegistry) ApplyChanges(ctx context.Context, changes *plan.Changes)
 		r.Labels[endpoint.OwnerLabelKey] = im.ownerID
 
-		filteredChanges.Create = append(filteredChanges.Create, im.generateTXTRecord(r)...)
+		generatedTXTs := im.generateTXTRecord(r)
+		for _, txt := range generatedTXTs {
+			// If the TXT record is already managed by this instance, skip it
+			if im.existingTXTs.isManaged(txt) {
+				continue
+			}
+			filteredChanges.Create = append(filteredChanges.Create, txt)
+		}
```
Thanks for the suggestion!

The reason the logic is structured this way is because `filteredChanges.Create` does not initially contain TXT records.  
The TXT records are generated on the fly using `im.generateTXTRecord(r)`, and those are what I intend to filter using `existingTXTs`.

Checking `r` directly (as in your example) wouldn’t work in this case, since it’s typically an A or CNAME record, not the TXT itself.  
Let me know if I misunderstood your suggestion!

u-kai replied:
Just pushed a refactor — the generateTXTRecordWithFilter part is a bit more involved, but the main loop is now cleaner and should be more efficient overall.
Open to feedback, let me know what you think!

u-kai commented:
@mloiseleur @ivankatliarchuk 

While working on the test implementation, I noticed a potential issue with the current behavior.

The current implementation correctly handles the case where an A record is accidentally deleted — it ensures that the A record is recreated. However, I realized that if all associated TXT records are manually deleted (while the corresponding A record remains), ExternalDNS currently has no way to detect and restore the missing TXT records. This is because the ownership information exists only in the TXT records, and the corresponding A record is not modified, so it does not appear in the change plan.

I’m currently considering two possible approaches to address this, and would appreciate any feedback or suggestions on which direction to take.

---

### Option 1: Extend `plan.Changes` with a `Desired` field

Currently, the `plan.Changes` struct only contains information about records that are being created, deleted, or updated (`Create`, `Delete`, `UpdateOld`, and `UpdateNew`).

This poses a limitation in scenarios where a TXT record has been manually deleted, but the corresponding A record still exists in the DNS provider. Since the A record is not being modified, it does not appear in the `Changes` structure — and as a result, `ApplyChanges` has no way to infer that a related TXT record should be recreated.

To address this, I’m proposing to extend `plan.Changes` with an additional `Desired` field, similar to the one already present in `plan.Plan`. This field would represent the complete set of records that should exist, regardless of whether they’re being changed or not.

With access to this `Desired` list in `ApplyChanges`, we could compare it against `existingTXTs` to detect missing TXT records and regenerate them if needed. Since this data is already held in memory by the `plan.Plan`, I believe the additional overhead would be minimal.

That said, I'm currently still exploring this approach and not yet certain whether it will work reliably in practice.

---

### Option 2: Document the limitation

An alternative would be to document this as a known limitation — that if a TXT record is deleted while the corresponding A record remains, ExternalDNS will not be able to restore the TXT record. This would avoid complicating the logic, but leave recovery in such cases as a manual task.

---

I’d be grateful for any thoughts or recommendations you might have.  
Thanks in advance!

ivankatliarchuk commented:
For now, worth to create an issue to track the scenario where TXT registry records are deleted. This is a prime candidate for a follow-up pull request in my opinion. If you happy try to resolve it, make sense to address.

Currently, if these 'database' records are removed, we lack a disaster recovery mechanism. Ideally, if records exist without corresponding registry entries, we'd want a one-time import operation, perhaps using --policy=import.

There is no point trying to resolve all issues/limitations with TXT registry in single PR

u-kai commented:
@ivankatliarchuk 
> For now, worth to create an issue to track the scenario where TXT registry records are deleted. This is a prime candidate for a follow-up pull request in my opinion. If you happy try to resolve it, make sense to address.



Thank you for the suggestion — I appreciate your perspective.

I agree that this is better handled in a separate pull request.
I’d like to take a shot at implementing it, so I’ll go ahead and open a follow-up PR to explore this scenario further.

Thanks again for the feedback!

ivankatliarchuk commented on registry/txt.go:
```diff
@@ -234,13 +273,17 @@ func (im *TXTRegistry) Records(ctx context.Context) ([]*endpoint.Endpoint, error
 // depending on the newFormatOnly configuration. The old format is maintained for backwards
 // compatibility but can be disabled to reduce the number of DNS records.
 func (im *TXTRegistry) generateTXTRecord(r *endpoint.Endpoint) []*endpoint.Endpoint {
+	return im.generateTXTRecordWithFilter(r, func(ep *endpoint.Endpoint) bool { return true })
```
```
	return im.generateTXTRecordWithFilter(r, func(ep *endpoint.Endpoint) bool { return true })
```

result of the funciton is always true, what is the reason for this function?

ivankatliarchuk commented:
/label tide/merge-method-squash

u-kai commented:
@ivankatliarchuk 
> Any chance to re-run smoke tests, just with bit more fixtures https://github.com/kubernetes-sigs/external-dns/pull/5459#issuecomment-2907762199 not 1 but 10-20 at least. Maybe worth to review function and variables names, rest seems ok to me.

I've added more test cases as you recommended.
While re-running the scenario from #5459, I discovered a regression caused by a recent change to the filtering logic for generatedTXT. Specifically, SetIdentifier was being applied after filtering, which resulted in expected TXT records being incorrectly skipped.

I’ve fixed that in the last commit, and verified the behavior manually in a real environment.
Here’s a snippet showing that A records are correctly recreated when missing, while existing TXT records are reused as expected:

```
------------------Apply Ingress ------------------
...
DEBU[0061] Adding api.hoge.com. to zone hoge.com. [Id: /hostedzone/ZONEXXXXXX]
DEBU[0061] Adding api.hoge.com. to zone hoge.com. [Id: /hostedzone/ZONEXXXXXX]
DEBU[0061] Adding a-api.hoge.com. to zone hoge.com. [Id: /hostedzone/ZONEXXXXXX]
INFO[0061] Desired change: CREATE a-hoge.hoge.com TXT  profile=default zoneID=/hostedzone/ZONEXXXXXX zoneName=hoge.com.
INFO[0061] Desired change: CREATE api.hoge.com A  profile=default zoneID=/hostedzone/ZONEXXXXXX zoneName=hoge.com.
INFO[0061] Desired change: CREATE api.hoge.com TXT  profile=default zoneID=/hostedzone/ZONEXXXXXX zoneName=hoge.com.
INFO[0061] 3 record(s) were successfully updated         profile=default zoneID=/hostedzone/ZONEXXXXXX zoneName=hoge.com.
...
------------------Manually Deleted api.hoge.com A record ------------------
DEBU[0122] Adding api.hoge.com. to zone hoge.com. [Id: /hostedzone/ZONEXXXXXX]
INFO[0122] Desired change: CREATE hoge.hoge.com A  profile=default zoneID=/hostedzone/ZONEXXXXXX zoneName=hoge.com.
INFO[0123] 1 record(s) were successfully updated         profile=default zoneID=/hostedzone/ZONEXXXXXX zoneName=hoge.com.
...
------------------Next Reconciliation Loop ------------------
INFO[0244] All records are already up to date
```

u-kai commented on registry/txt.go:
```diff
@@ -60,6 +62,41 @@ type TXTRegistry struct {
+	im.entries[fmt.Sprintf("%s|%s", r.DNSName, r.SetIdentifier)] = struct{}{}
+}
+
+// filterOutExistingTXTRecords removes endpoints whose TXT companions are already present
```
Fix it!

ivankatliarchuk reviewed (approved):
/lgtm

ivankatliarchuk commented:
Check out this new issue I created detailing all the known TXT registry bugs: https://github.com/kubernetes-sigs/external-dns/issues/5465. It's worth a look – you might spot something interesting to tackle. Just a heads-up, not all of them have straightforward solutions or even have a solution at the moment.

szuecs commented on registry/txt.go:
```diff
@@ -60,6 +62,42 @@ type TXTRegistry struct {
+
+	// existingTXTs is the TXT records that already exist in the zone so that
+	// ApplyChanges() can skip re-creating them. See the struct below for details.
+	existingTXTs existingTXTs
```
why it's not a pointer type?

u-kai replied:
After thinking it through, I realized that many operations mutate the value, so using a pointer felt more natural.
I've updated the code to use a pointer accordingly.

szuecs commented:
> Thanks. I'm not sure of this approach 🤔 It may increase memory usage on big users. /ok-to-test cc @szuecs

I am a bit worried introducing a cache in this place. It's not really a memory issue, but we should not add state, if we do not need it. There is also no cache invalidation.
If I read https://github.com/kubernetes-sigs/external-dns/issues/4914 correctly, then likely the problem is a configuration issue as stated in https://github.com/kubernetes-sigs/external-dns/issues/4914#issuecomment-2520210855 .
Another common issue is that external-dns thinks that there is another owner, so "lost ownership problem".

/hold clarify if this fixes a problem

ivankatliarchuk commented:
relates https://github.com/kubernetes-sigs/external-dns/issues/3307

mloiseleur commented:
@u-kai If you rebase on top of master, the CI should go green.

u-kai commented:
One of the flaky tests failed. Looking at the test code, it waits up to 2 seconds for failCount to reach at least 3, so it likely failed due to timing. I will rerun the tests.

/retest

ivankatliarchuk commented:
/unset approved

mloiseleur commented:
/remove-approve

mloiseleur commented:
/approve cancel

szuecs commented:
> 
> Also, when only the TXT record is deleted, it currently won’t be recreated — this is a separate issue that I plan to address in a future PR, also using `existingTXTs`.

We don't want to recreate TXT records in this case!
It's the mechanism that identifies the ownership and if you mix a zone with multiple external-dns or external-dns and manual records you would create a mess if you would recreate the deleted txt record. A txt record deletion means, that you don't want to let this record being owned by the external-dns instance.

szuecs commented on registry/txt.go:
```diff
@@ -244,7 +301,9 @@ func (im *TXTRegistry) generateTXTRecord(r *endpoint.Endpoint) []*endpoint.Endpo
 			txt.Labels[endpoint.OwnedRecordLabelKey] = r.DNSName
 			txt.ProviderSpecific = r.ProviderSpecific
-			endpoints = append(endpoints, txt)
+			if filter(txt) {
```
That's basically `if true` as far as I understand https://github.com/kubernetes-sigs/external-dns/pull/5459/files#r2107584832 , right?
If so please drop the added complexity. If we need filtering than have it but other than that I don't think we should have an `if` branch here.

u-kai replied:
I think we still need the if here because we use isNotManaged for filtering.
The purpose of this change is to filter out existing TXT records when an A record was mistakenly deleted and later recreated by External-DNS.
This filtering is necessary because trying to create a TXT record that already exists would result in an error. So if a TXT record already exists, we avoid creating it again.

szuecs commented on registry/txt.go:
```diff
@@ -279,7 +342,7 @@ func (im *TXTRegistry) ApplyChanges(ctx context.Context, changes *plan.Changes)
 		r.Labels[endpoint.OwnerLabelKey] = im.ownerID
 
-		filteredChanges.Create = append(filteredChanges.Create, im.generateTXTRecord(r)...)
+		filteredChanges.Create = append(filteredChanges.Create, im.generateTXTRecordWithFilter(r, im.existingTXTs.isNotManaged)...)
```
I am not sure if this is correct

u-kai replied:
see hear 🙇 https://github.com/kubernetes-sigs/external-dns/pull/5459#discussion_r2144854049

szuecs reviewed (changes requested):
A bit of cleanup and missing negative tests

For sure we do not want to recreate ownership records.

u-kai commented on registry/txt.go:
```diff
@@ -60,6 +61,56 @@ type TXTRegistry struct {
+}
+
+// isNotManaged reports whether the given endpoint's TXT record is absent from the existing set.
+// Used to determine whether a new TXT record needs to be created.
```
What I meant here was that this is used to determine whether a TXT record should be created — I didn’t intend to say that it is about recreating the record.
But if this is confusing, I’m happy to remove it.

szuecs replied:
I see but I think it's more confusing to have this comment.
I also see that how it's integrated we store only TXT records. This filters other records before `*existingTXTs.add()` is called https://github.com/kubernetes-sigs/external-dns/pull/5459/files#diff-353145d31754c56c5c08d53f4a5b1ee4cf2116cd605bbb9a7bc572c9072b2ab8R192-R194

I think in general the function name should be `isAbsent` or `isNotKnown` or `isUnknown` or `isNotCached`. We can think of `existingTXTs.entries` as an in-memory cache/store.
It's used as `im.generateTXTRecordWithFilter(r, im.existingTXTs.isNotManaged)`

```
im.generateTXTRecordWithFilter(r, im.existingTXTs.isAbsent)
im.generateTXTRecordWithFilter(r, im.existingTXTs.isNotKnown)
im.generateTXTRecordWithFilter(r, im.existingTXTs.isUnknown)
im.generateTXTRecordWithFilter(r, im.existingTXTs.isNotCached)
```

For me `im.generateTXTRecordWithFilter(r, im.existingTXTs.isUnknown)` reads best.
wdyt?

u-kai replied:
Thanks for the review!
I renamed the helper to `isAbsent` to make it clear that `ExistingTXTs` filters out records that already exist; we only create when a record is absent.

u-kai commented on registry/txt.go:
```diff
@@ -259,14 +318,18 @@ func (im *TXTRegistry) generateTXTRecord(r *endpoint.Endpoint) []*endpoint.Endpo
 		txtNew.Labels[endpoint.OwnedRecordLabelKey] = r.DNSName
 		txtNew.ProviderSpecific = r.ProviderSpecific
-		endpoints = append(endpoints, txtNew)
+		if filter(txtNew) {
```
see hear 🙇 https://github.com/kubernetes-sigs/external-dns/pull/5459#discussion_r2144854049

ivankatliarchuk commented on registry/txt.go:
```diff
@@ -58,6 +59,56 @@ type TXTRegistry struct {
+	}
+}
+
+func (im *existingTXTs) add(r *endpoint.Endpoint) {
```
function not fully covered

![Screenshot 2025-07-03 at 08 01 10](https://github.com/user-attachments/assets/d44db340-d752-4f27-b273-de38d1ba5003)

ivankatliarchuk replied:
But agree, we need to be super careful, as pretty much every deployment rely on txt registry

ivankatliarchuk reviewed (changes requested):
rest seems legit to me

We just w8 for @szuecs view

szuecs commented:
> However, if the goal is to stop managing certain records, wouldn't it be more appropriate to configure domainFilter or similar mechanisms? What do you think about that approach?

No that shouldn't be the tool.
Txt registry is the fundamental ownership record such that you can have multiple clusters with multiple external-dns and whatever for example OpenTofu to manage a single zone.
If we rewrite txt records we should do this only in a script or run-once in order to hand over records from one external-dns to another one or rewrite prefix/suffix patterns to be able to change the design of your infrastructure.

u-kai commented on registry/txt_test.go:
```diff
@@ -1838,3 +1887,201 @@ func TestTXTRegistryRecordsWithEmptyTargets(t *testing.T) {
+				newEndpointWithOwner("record-1.test-zone.example.org", "1.1.1.1", endpoint.RecordTypeA, ownerId),
+			},
+		},
+		// TODO: Test TXT record regeneration when only A/CNAME records exist.
```
I’ve removed it.

u-kai commented:
@szuecs 
Thanks for the clarification — I now understand the role of the TXT record and that it shouldn't be recreated.
I've made the changes as you suggested. I'd appreciate it if you could take another look.

szuecs commented on registry/txt_test.go:
```diff
@@ -1838,3 +1887,201 @@ func TestTXTRegistryRecordsWithEmptyTargets(t *testing.T) {
+			expectedCreate: []*endpoint.Endpoint{
+				newEndpointWithOwner("record-1.test-zone.example.org", "1.1.1.1", endpoint.RecordTypeA, ownerId),
+			},
+		},
```
You could have 2 clusters working on the same zone and competing ingress for the same hostname for example.
The "no owner" case is if I add manually a record `foo.example` and later I create an ingress with `foo.example` host, then external-dns should not change the created record.

I want to make sure we do not break ownership here.

I think you created the tests as requested below.

szuecs reviewed (commented):
Some non critical changes I would like to see.

Thanks for your work!

u-kai commented:
@szuecs 
Thank you for the review. I've applied the requested changes. If there are no problems, I'd appreciate your approval. Please take a look when you have a moment.

szuecs commented:
/approve

ivankatliarchuk commented:
/hold cancel
